compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+                .foldLeft((0,0)) { case ((errors, warnings), reporter) =>
+                  if (reporter.errorCount > 0) {
+                    val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+                    addFailureInstruction(errorMsg)
+                    reporter.echo(errorMsg)
+                    reporter.flushToFile()
+                  }
+
+                  (errors + reporter.errorCount, warnings + reporter.warningCount)
+                }
+
+            if (errorCount > 0) fail()
+
+            completeCompilation(errorCount)
+            (errorCount, warningCount, checkFile.exists, () => verifyOutput(checkFile, outDir, target, warningCount))
+          }
+        }
+
+        if (errorCount == 0 && hasCheckFile) doVerify()
+        else if (errorCount > 0) {
+          echo(s"\nCompilation failed for: '$target'")
+          val buildInstr = target.buildInstructions(errorCount, warningCount)
+          addFailureInstruction(buildInstr)
+          failTarget(target)
+        }
+      }
+    }
+  }
+
+  private final class NegTest(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean)
+  extends Test(targets, times, threadLimit, suppressAllOutput) {
+    protected def compilationRunnable(target: Target): Runnable = new Runnable {
+      def run(): Unit = compileTry {
+        // In neg-tests we allow two types of error annotations,
+        // "nopos-error" which doesn't care about position and "error" which
+        // has to be annotated on the correct line number.
+        //
+        // We collect these in a map `"file:row" -> numberOfErrors`, for
+        // nopos errors we save them in `"file" -> numberOfNoPosErrors`
+        def errorMapAndExpected(files: Array[JFile]): (HashMap[String, Integer], Int) = {
+          val errorMap = new HashMap[String, Integer]()
+          var expectedErrors = 0
+          files.filter(_.getName.endsWith(".scala")).foreach { file =>
+            Source.fromFile(file).getLines.zipWithIndex.foreach { case (line, lineNbr) =>
+              val errors = line.sliding("// error".length).count(_.mkString == "// error")
+              if (errors > 0)
+                errorMap.put(s"${file.getAbsolutePath}:${lineNbr}", errors)
+
+              val noposErrors = line.sliding("// nopos-error".length).count(_.mkString == "// nopos-error")
+              if (noposErrors > 0) {
+                val nopos = errorMap.get("nopos")
+                val existing: Integer = if (nopos eq null) 0 else nopos
+                errorMap.put("nopos", noposErrors + existing)
+              }
+
+              expectedErrors += noposErrors + errors
+            }
+          }
+
+          (errorMap, expectedErrors)
+        }
+
+        def getMissingAnnotations(errorMap: HashMap[String, Integer], reporterErrors: Iterator[MessageContainer]) = !reporterErrors.forall { error =>
```

Comment:
Indicators is also the wrong term - annotations is the generic term, `@-annotations` are specific to what you're referring to.

Hmm, I wonder what synonym to use for this - perhaps `getMissingExpectedErrors`?